cloudflare/resource_cloudflare_authenticated_origin_pulls.go

```diff
@@ -0,0 +1,150 @@
+package cloudflare
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/pkg/errors"
+)
+
+func resourceCloudflareAuthenticatedOriginPulls() *schema.Resource {
+	return &schema.Resource{
+		// AOP is a toggleable feature, editing is the same as creating.
```

Comment:
Correct, the only state that you can edit on AOP is whether it is enabled or not. The additional parameter of which AOP you'd like to use I've abstracted based on the inputs provided.

One exception I'd like to call out to this is Per Hostname AOP which the v4 API only exposes through one endpoint that we can use, `EditPerHostnameAuthenticatedOriginPullsConfig`, as shown below from the `Create` function:

```
...
	case hostname != "" && aopCert != "":
		// Per Hostname AOP
		conf := []cloudflare.PerHostnameAuthenticatedOriginPullsConfig{{
			CertID:   aopCert,
			Hostname: hostname,
			Enabled:  isEnabled.(bool),
		}}
		_, err := client.EditPerHostnameAuthenticatedOriginPullsConfig(zoneID, conf)
		if err != nil {
			return fmt.Errorf("Error creating Per-Hostname Authenticated Origin Pulls resource on zone %q: %s", zoneID, err)
		}
		checksum := stringChecksum(fmt.Sprintf("PerHostnameAOP/%s/%s/%s", zoneID, hostname, aopCert))
		d.SetId(checksum)
...
```

Here, we may be `PUT`ing a new config over an existing one, or placing a new one in altogether. I decided to not write an entire `Edit` function to cover this specific use case because the code would not only be the duplicate of above because creating/editing are the same endpoint but also that per-zone and global would need to be brought over in addition. 

As for `authenticated_origin_pulls_certificate`, they may only be created or deleted.

Reply:
Thanks for the explanation! It's all alot cleaner now 🙂